File: packages/python/src/agent_wallet/secret/kv_store.py
"""Secret layer: Keystore V3 encryption engine for private keys and credentials."""

from __future__ import annotations

import json
import os
from pathlib import Path

from Crypto.Cipher import AES
from Crypto.Protocol.KDF import scrypt as scrypt_kdf
from eth_hash.auto import keccak

from agent_wallet.core.errors import DecryptionError

# Keystore V3 scrypt parameters
SCRYPT_N = 262144
SCRYPT_R = 8
SCRYPT_P = 1
SCRYPT_DKLEN = 32

# Sentinel value for master.json password verification
MASTER_SENTINEL = b"agent-wallet"


def _derive_key(password: str, salt: bytes) -> bytes:
    """Derive a 32-byte key from password + salt using scrypt."""
    return scrypt_kdf(
        password.encode("utf-8"),
        salt,
        key_len=SCRYPT_DKLEN,
        N=SCRYPT_N,
        r=SCRYPT_R,
        p=SCRYPT_P,
    )


def _encrypt_bytes(plaintext: bytes, password: str) -> dict:
    """Encrypt arbitrary bytes using Keystore V3 compatible format.

    Returns a JSON-serializable dict with Keystore V3 structure.
    """
    salt = os.urandom(32)
    iv = os.urandom(16)
    derived_key = _derive_key(password, salt)

    # AES-128-CTR: use first 16 bytes of derived key
    encryption_key = derived_key[:16]
    cipher = AES.new(encryption_key, AES.MODE_CTR, nonce=b"", initial_value=iv)
    ciphertext = cipher.encrypt(plaintext)

    # MAC: keccak256(mac_key + ciphertext)
    mac_key = derived_key[16:]
    mac = keccak(mac_key + ciphertext)

    return {
        "version": 3,
        "crypto": {
            "cipher": "aes-128-ctr",
            "cipherparams": {"iv": iv.hex()},
            "ciphertext": ciphertext.hex(),
            "kdf": "scrypt",
            "kdfparams": {
                "dklen": SCRYPT_DKLEN,
                "n": SCRYPT_N,
                "r": SCRYPT_R,
                "p": SCRYPT_P,
                "salt": salt.hex(),
            },
            "mac": mac.hex(),
        },
    }


def _decrypt_bytes(keystore: dict, password: str) -> bytes:
    """Decrypt a Keystore V3 compatible JSON dict back to plaintext bytes."""
    crypto = keystore["crypto"]
    kdfparams = crypto["kdfparams"]

    salt = bytes.fromhex(kdfparams["salt"])
    iv = bytes.fromhex(crypto["cipherparams"]["iv"])
    ciphertext = bytes.fromhex(crypto["ciphertext"])
    stored_mac = crypto["mac"]

    derived_key = _derive_key(password, salt)

    # Verify MAC before decrypting
    mac_key = derived_key[16:]
    computed_mac = keccak(mac_key + ciphertext).hex()
    if computed_mac != stored_mac:
        raise DecryptionError("MAC mismatch — wrong password or corrupted file")

    # Decrypt
    encryption_key = derived_key[:16]
    cipher = AES.new(encryption_key, AES.MODE_CTR, nonce=b"", initial_value=iv)
    return cipher.decrypt(ciphertext)


class SecureKVStore:
    """Keystore V3 based encryption engine for keys and credentials.

    Holds the password for the duration of its lifetime (typically only during
    WalletRegistry.__init__). After WalletRegistry init completes, both the
    password and this KVStore instance go out of scope.
    """

    def __init__(self, secrets_dir: str | Path, password: str) -> None:
        self._secrets_dir = Path(secrets_dir)
        self._password = password

        if not self._secrets_dir.is_dir():
            raise FileNotFoundError(f"Secrets directory not found: {self._secrets_dir}")

    # --- Master Password ---

    def init_master(self) -> None:
        """Create master.json sentinel file (called by `agent-wallet init`)."""
        keystore = _encrypt_bytes(MASTER_SENTINEL, self._password)
        self._write_json("master.json", keystore)

    def verify_password(self) -> bool:
        """Verify password against master.json sentinel.

        Raises DecryptionError if password is wrong.
        Returns True on success.
        """
        path = self._secrets_dir / "master.json"
        if not path.exists():
            raise FileNotFoundError(
                "master.json not found. Run `agent-wallet init` first."
            )
        keystore = self._read_json("master.json")
        plaintext = _decrypt_bytes(keystore, self._password)
        if plaintext != MASTER_SENTINEL:
            raise DecryptionError("master.json decrypted but sentinel mismatch")
        return True

    # --- Identity (private keys, fixed 32 bytes) ---

    def load_private_key(self, name: str) -> bytes:
        """Decrypt id_<name>.json and return raw 32-byte private key."""
        filename = f"id_{name}.json"
        keystore = self._read_json(filename)
        return _decrypt_bytes(keystore, self._password)

    def save_private_key(self, name: str, private_key: bytes) -> None:
        """Encrypt and save a 32-byte private key as id_<name>.json."""
        if len(private_key) != 32:
            raise ValueError(f"Private key must be 32 bytes, got {len(private_key)}")
        keystore = _encrypt_bytes(private_key, self._password)
        self._write_json(f"id_{name}.json", keystore)

    def generate_key(self, name: str) -> bytes:
        """Generate a random 32-byte private key, save it, return the key."""
        private_key = os.urandom(32)
        self.save_private_key(name, private_key)
        return private_key

    # --- Credentials (arbitrary length) ---

    def load_credential(self, name: str) -> str | dict:
        """Decrypt cred_<name>.json and return the credential value."""
        filename = f"cred_{name}.json"
        keystore = self._read_json(filename)
        plaintext = _decrypt_bytes(keystore, self._password)
        text = plaintext.decode("utf-8")
        # Try to parse as JSON dict (for structured credentials)
        try:
            parsed = json.loads(text)
            if isinstance(parsed, dict):
                return parsed
        except json.JSONDecodeError:
            pass
        return text

    def save_credential(self, name: str, value: str | dict) -> None:
        """Encrypt and save a credential as cred_<name>.json."""
        if isinstance(value, dict):
            plaintext = json.dumps(value, ensure_ascii=False).encode("utf-8")
        else:
            plaintext = value.encode("utf-8")
        keystore = _encrypt_bytes(plaintext, self._password)
        self._write_json(f"cred_{name}.json", keystore)

    # --- Internal helpers ---

    def _read_json(self, filename: str) -> dict:
        path = self._secrets_dir / filename
        if not path.exists():
            raise FileNotFoundError(f"Keystore file not found: {path}")
        return json.loads(path.read_text(encoding="utf-8"))

    def _write_json(self, filename: str, data: dict) -> None:
        path = self._secrets_dir / filename
        path.write_text(
            json.dumps(data, indent=2, ensure_ascii=False) + "\n",
            encoding="utf-8",
        )

File: packages/typescript/src/local/kv-store.ts
import { createCipheriv, createDecipheriv, randomBytes, scryptSync } from "node:crypto";
import { readFileSync, writeFileSync, existsSync, statSync, chmodSync } from "node:fs";
import { join } from "node:path";
import { keccak256 } from "viem";
import { DecryptionError } from "../core/errors.js";

const SCRYPT_N = 262144;
const SCRYPT_R = 8;
const SCRYPT_P = 1;
const SCRYPT_DKLEN = 32;

const MASTER_SENTINEL = Buffer.from("agent-wallet", "utf-8");

interface KeystoreV3 {
  version: 3;
  crypto: {
    cipher: string;
    cipherparams: { iv: string };
    ciphertext: string;
    kdf: string;
    kdfparams: {
      dklen: number;
      n: number;
      r: number;
      p: number;
      salt: string;
    };
    mac: string;
  };
}

function deriveKey(password: string, salt: Uint8Array): Buffer {
  return scryptSync(Buffer.from(password, "utf-8"), salt, SCRYPT_DKLEN, {
    N: SCRYPT_N,
    r: SCRYPT_R,
    p: SCRYPT_P,
    maxmem: 256 * SCRYPT_N * SCRYPT_R,
  }) as Buffer;
}

export function encryptBytes(plaintext: Uint8Array, password: string): KeystoreV3 {
  const salt = randomBytes(32);
  const iv = randomBytes(16);
  const derivedKey = deriveKey(password, salt);

  const encryptionKey = derivedKey.subarray(0, 16);
  const cipher = createCipheriv("aes-128-ctr", encryptionKey, iv);
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);

  const macKey = derivedKey.subarray(16);
  const macInput = Buffer.concat([macKey, ciphertext]);
  const mac = keccak256(macInput).slice(2); // strip 0x

  return {
    version: 3,
    crypto: {
      cipher: "aes-128-ctr",
      cipherparams: { iv: iv.toString("hex") },
      ciphertext: ciphertext.toString("hex"),
      kdf: "scrypt",
      kdfparams: {
        dklen: SCRYPT_DKLEN,
        n: SCRYPT_N,
        r: SCRYPT_R,
        p: SCRYPT_P,
        salt: salt.toString("hex"),
      },
      mac,
    },
  };
}

export function decryptBytes(keystore: KeystoreV3, password: string): Uint8Array {
  const { crypto } = keystore;
  const { kdfparams } = crypto;

  const salt = Buffer.from(kdfparams.salt, "hex");
  const iv = Buffer.from(crypto.cipherparams.iv, "hex");
  const ciphertext = Buffer.from(crypto.ciphertext, "hex");
  const storedMac = crypto.mac;

  const derivedKey = deriveKey(password, salt);

  const macKey = derivedKey.subarray(16);
  const macInput = Buffer.concat([macKey, ciphertext]);
  const computedMac = keccak256(macInput).slice(2);

  if (computedMac !== storedMac) {
    throw new DecryptionError("MAC mismatch — wrong password or corrupted file");
  }

  const encryptionKey = derivedKey.subarray(0, 16);
  const decipher = createDecipheriv("aes-128-ctr", encryptionKey, iv);
  return Uint8Array.from(Buffer.concat([decipher.update(ciphertext), decipher.final()]));
}

export class SecureKVStore {
  private secretsDir: string;
  private password: string;

  constructor(secretsDir: string, password: string) {
    this.secretsDir = secretsDir;
    this.password = password;

    if (!existsSync(secretsDir) || !statSync(secretsDir).isDirectory()) {
      throw new Error(`Secrets directory not found: ${secretsDir}`);
    }
  }

  initMaster(): void {
    const keystore = encryptBytes(MASTER_SENTINEL, this.password);
    this.writeJson("master.json", keystore);
  }

  verifyPassword(): boolean {
    const masterPath = join(this.secretsDir, "master.json");
    if (!existsSync(masterPath)) {
      throw new Error("master.json not found. Run `agent-wallet init` first.");
    }
    const keystore = this.readJson("master.json") as KeystoreV3;
    const plaintext = decryptBytes(keystore, this.password);
    if (Buffer.compare(Buffer.from(plaintext), MASTER_SENTINEL) !== 0) {
      throw new DecryptionError("master.json decrypted but sentinel mismatch");
    }
    return true;
  }

  loadSecret(name: string): Uint8Array {
    const keystore = this.readJson(`secret_${name}.json`) as KeystoreV3;
    return decryptBytes(keystore, this.password);
  }

  saveSecret(name: string, secret: Uint8Array): void {
    const keystore = encryptBytes(secret, this.password);
    this.writeJson(`secret_${name}.json`, keystore);
  }

  generateSecret(name: string, opts?: { length?: number }): Uint8Array {
    const secret = randomBytes(opts?.length ?? 32);
    this.saveSecret(name, secret);
    return Uint8Array.from(secret);
  }

  loadCredential(name: string): string | Record<string, unknown> {
    const keystore = this.readJson(`cred_${name}.json`) as KeystoreV3;
    const plaintext = decryptBytes(keystore, this.password);
    const text = Buffer.from(plaintext).toString("utf-8");
    try {
      const parsed = JSON.parse(text);
      if (typeof parsed === "object" && parsed !== null && !Array.isArray(parsed)) {
        return parsed as Record<string, unknown>;
      }
    } catch {
      // not JSON, return raw text
    }
    return text;
  }

  saveCredential(name: string, value: string | Record<string, unknown>): void {
    const plaintext =
      typeof value === "string"
        ? Buffer.from(value, "utf-8")
        : Buffer.from(JSON.stringify(value), "utf-8");
    const keystore = encryptBytes(plaintext, this.password);
    this.writeJson(`cred_${name}.json`, keystore);
  }

  private readJson(filename: string): unknown {
    const path = join(this.secretsDir, filename);
    if (!existsSync(path)) {
      throw new Error(`Keystore file not found: ${path}`);
    }
    return JSON.parse(readFileSync(path, "utf-8"));
  }

  private writeJson(filename: string, data: unknown): void {
    const path = join(this.secretsDir, filename);
    writeFileSync(path, JSON.stringify(data, null, 2) + "\n", "utf-8");
    try {
      chmodSync(path, 0o600);
    } catch {
      // ignore on platforms without chmod support
    }
  }
}
